2023/Q1/artifacts/PoS-pseudocode/PoS-model-redelegation.md

```diff
@@ -636,13 +624,10 @@ end_of_epoch()
       var total_staked = read_epoched_field(validators[validator_address].total_deltas, slash.epoch, 0)
       slash_misbehaving_validator(validator_address, slash, total_staked)
 
-      forall (val in validators) do
-        var pairs = { pair = <delegator, <epoch, redelegation> | pair in validators[val].redelegations[validator_address] &&
-                                                                 pair !=⊥ && 
-                                                                 redelegation.epoch - unbonding_length <= slash.epoch < redelegation.epoch + pipeline_length}
-        forall (<delegator, redelegation> in pairs) do
-          var staked_redelegated = 
-          slash_redelegated_validator(val, validator_address, slash, redelegation.amount)
+      forall (dest_validators in validators[validator_address].outgoing_redelegations.keys()) do
+        forall (amount in {amount | amount = validators[validator_address].outgoing_redelegations[dest_validators][start] > 0 &&
```

Comment:
I think there's a slight difference here from the previous version - the `start` epoch is not defined in this scope. We probably want to iter over the outgoing redelegations for each `validator_address, dest_validators` pair and use the same condition as before for their epochs, i.e. `redelegation.epoch - unbonding_length <= slash.epoch < redelegation.epoch + pipeline_length`

Reply:
That's the intention of the second condition in the line below: `start - unbonding_length <= slash.epoch < start + pipeline_length`. Maybe it is not well expressed though, we can always make the second iteration more explicit.